timemachine/cpp/src/langevin_integrator.cu

```diff
@@ -0,0 +1,93 @@
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "langevin_integrator.hpp"
+#include "math_utils.cuh"
+
+#include "kernels/k_integrator.cuh"
+
+namespace timemachine {
+
+int round_up_even(int count) {
+    if (count % 2 == 1) {
+        return count += 1;
+    } else {
+        return count;
+    }
+}
+
+LangevinIntegrator::LangevinIntegrator(int N, double dt, double ca, const double *h_cbs, const double *h_ccs, int seed)
+    : N_(N), dt_(dt), ca_(ca) {
+
+    d_cbs_ = gpuErrchkCudaMallocAndCopy(h_cbs, N);
+    d_ccs_ = gpuErrchkCudaMallocAndCopy(h_ccs, N);
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+    gpuErrchk(cudaMalloc(&d_noise_, round_up_even(N * 3) * sizeof(double)));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+
+    gpuErrchk(cudaMalloc(&d_du_dx_, N * 3 * sizeof(*d_du_dx_)));
+}
+
+LangevinIntegrator::~LangevinIntegrator() {
+    gpuErrchk(cudaFree(d_cbs_));
+    gpuErrchk(cudaFree(d_ccs_));
+    gpuErrchk(cudaFree(d_noise_));
+    gpuErrchk(cudaFree(d_du_dx_));
+    curandErrchk(curandDestroyGenerator(cr_rng_));
+}
+
+void LangevinIntegrator::step_fwd(
+    std::vector<BoundPotential *> &bps,
```

Comment:
can you remind me again why we're shifting the responsibility of the force calculation to the `step()` function as opopsed to the context? (i.e. why can't we just pass in a `double *d_du_dx_t` instead?

Reply:
(and similar for `initialize()` and `finalize()`)

Reply:
in the old behavior, the context sometimes will compute multiple values (eg. `U`, `dU_dx`, `dU_dp`, `dU_dl`) in a single pass, where as here, you will force the integrator to compute an exra `dU_dx` on every step.

Reply:
In part it is more consistent with our reference implementations, and it also allows the integrator to compute the forces as many times as it would like. Which seems like a cleaner separation of responsibilities.

Currently the only time we compute additional values at the same time as `dU_dx` is `dU_dl` which this has a carve out for by passing in a buffer. The other case where we compute values on frames (`multiple_steps_U`) is already a separate potential call to the force calculation.

Reply:
hm okay - I see! I'm okay with this.